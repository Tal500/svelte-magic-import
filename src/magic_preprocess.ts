import MagicString from 'magic-string';

export const magic_preprocess = () => ({
	markup: ({ content, filename }: { content: string; filename: string }) => {
		console.log('this runs first');
		//console.log('content: ' + content);
		console.log('filename: ' + filename);
		const result = content.match(/<script.*?<\/script>/gs); // Or match all?
		//console.log("Result: " + JSON.stringify(result));

		const s: MagicString = new MagicString(content, { filename }); // TODO: Use it

		if (result) {
			//console.log("Found0: " + result[0])
			const importStatement = result[0].match(/import { shared } from '\$lib';/gs);
			if (importStatement) {
				// TODO: Use magic string
				console.log('Found');
				s.replace(importStatement[0], '');

				// TODO: Strip spaces inside the parameter, and verify that the word inside is a legal one.
				s.replaceAll(/::shared\(some\)/gs, (/*match, offset*/) => ':global(.svelte-magic-blabla)'); // TODO: Only on style
				s.replaceAll(/(?<!:):shared\(some\)/gs, (/*match, offset*/) => '.svelte-magic-blabla'); // TODO: Only on style
				s.replaceAll(/(?<!:)shared\('some'\)/gs, (/*match, offset*/) => '"svelte-magic-blabla"'); // TODO: Only outside style

				console.log('Modified content: ' + s.toString());
			} else {
				console.log('Not Found');
			}
		}

		return {
			code: s.toString(),
			map: s.generateMap()
		};
	}
});
