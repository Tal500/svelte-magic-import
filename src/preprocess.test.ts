import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, test, expect } from 'vitest';

import magic_preprocess from './preprocess.js';

const __dirname = resolve(fileURLToPath(import.meta.url), '..');

const readInputFile = (filename: string) =>
  readFileSync(resolve(__dirname, 'test-input', filename), {
    encoding: 'utf8',
    flag: 'r'
  });

const snippets = [
  'simple.svelte'
];

describe('preprocess snapshots', () => {
    snippets.forEach((snippet) => {
        test(snippet, () => {
          expect(
            magic_preprocess().markup({ content: readInputFile(snippet), filename: snippet })
          ).toMatchSnapshot()
        })
      });
});

// TODO: Write better tests instead of snapshots:
//  Check that it switched the correct places of definitions