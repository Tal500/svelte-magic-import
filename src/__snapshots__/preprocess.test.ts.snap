// Vitest Snapshot v1

exports[`preprocess snapshots > simple.svelte 1`] = `
{
  "code": "<script lang=\\"ts\\">
	
</script>

<h1 class={\\"svelte-magic-some\\"}>Welcome to your library project</h1>
<p>Create your package using @sveltejs/package and preview/showcase your work with SvelteKit</p>
<p>Visit <a href=\\"https://kit.svelte.dev\\">kit.svelte.dev</a> to read the documentation</p>

<style>
	:global(.svelte-magic-some) {
		color: red;
	}
</style>
",
  "map": SourceMap {
    "file": null,
    "mappings": "AAAA;AACA,CAAkD;AAClD;AACA;AACA,WAAW,mBAAc;AACzB;AACA;AACA;AACA;AACA,CAAC,2BAAc;AACf;AACA;AACA;",
    "names": [],
    "sources": [
      null,
    ],
    "sourcesContent": [
      null,
    ],
    "version": 3,
  },
}
`;
